File: app/src/main/java/mega/privacy/android/app/utils/StringResourcesUtils.java
package mega.privacy.android.app.utils;

import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;

import java.util.Locale;

import mega.privacy.android.app.MegaApplication;
import mega.privacy.android.app.R;
import nz.mega.sdk.MegaChatError;
import nz.mega.sdk.MegaError;

import static mega.privacy.android.app.utils.LogUtil.*;
import static nz.mega.sdk.MegaChatError.*;
import static nz.mega.sdk.MegaError.*;

public class StringResourcesUtils {

    // English locale resources
    private static Resources enRes;

    /**
     * Get the English locale resources
     * @param context The current context.
     * @return The English locale resources.
     */
    private static Resources getENResource(Context context) {
        if (enRes == null) {
            Configuration configuration = new Configuration(context.getResources().getConfiguration());
            configuration.setLocale(new Locale(Locale.ENGLISH.getLanguage()));
            enRes = context.createConfigurationContext(configuration).getResources();
        }
        return enRes;
    }

    /**
     * Return the string value associated with a particular resource ID in the current app language.
     * If something fails return the original string in English.
     *
     * @param resId The resource identifier of the desired string.
     * @return The desired string.
     */
    public static String getString(int resId) {

        MegaApplication app = MegaApplication.getInstance();

        try {
            return app.getString(resId);
        } catch (Exception e) {
            logWarning("Error getting a translated string.", e);
            String originalString = getENResource(app).getString(resId);
            logInfo("Using the original English string: " + originalString);
            return originalString;
        }
    }

    /**
     * Return the string value associated with a particular resource ID in the current app language,
     * substituting the format arguments as defined in Formatter and String.format(String, Object...).
     * If something fails return the original string in English.
     *
     * @param resId      The resource identifier of the desired string.
     * @param formatArgs The format arguments that will be used for substitution.
     * @return The desired string.
     */
    public static String getString(int resId, Object... formatArgs) {

        MegaApplication app = MegaApplication.getInstance();

        try {
            return app.getString(resId, formatArgs);
        } catch (Exception e) {
            logWarning("Error getting a translated and formatted string.", e);
            String originalString = getENResource(app).getString(resId, formatArgs);
            logInfo("Using the original English string: " + originalString);
            return originalString;
        }
    }

    /**
     * Returns the string necessary for grammatically correct pluralization of the given
     * resource ID for the given quantity in the current app language.
     * If something fails return the original string in English.
     *
     * @param resId    The resource identifier of the desired string
     * @param quantity The number used to get the correct string for the current language's plural rules.
     * @return The desired string.
     */
    public static String getQuantityString(int resId, int quantity) {

        MegaApplication app = MegaApplication.getInstance();

        try {
            return app.getResources().getQuantityString(resId, quantity);
        } catch (Exception e) {
            logWarning("Error getting a translated string with quantity modifier.", e);
            String originalString = getENResource(app).getQuantityString(resId, quantity);
            logInfo("Using the original English string: " + originalString);
            return originalString;
        }
    }

    /**
     * Returns the string necessary for grammatically correct pluralization of the given resource ID
     * for the given quantity in the current app language and substituting the format arguments as
     * defined in Formatter and String.format(String, Object...).
     * If something fails return the original string in English.
     *
     * @param resId      The resource identifier of the desired string.
     * @param quantity   The number used to get the correct string for the current language's plural rules.
     * @param formatArgs The format arguments that will be used for substitution.
     * @return The desired string.
     */
    public static String getQuantityString(int resId, int quantity, Object... formatArgs) {

        MegaApplication app = MegaApplication.getInstance();

        try {
            return app.getResources().getQuantityString(resId, quantity, formatArgs);
        } catch (Exception e) {
            logWarning("Error getting a translated and formatted string with quantity modifier.", e);
            String originalString = getENResource(app).getQuantityString(resId, quantity, formatArgs);
            logInfo("Using the original English string: " + originalString);
            return originalString;
        }
    }

    /**
     * Gets the translated string of an error received in a request.
     *
     * @param error MegaError received in the request
     * @return The translated string
     */
    public static String getTranslatedErrorString(MegaError error) {
        MegaApplication app = MegaApplication.getInstance();
        if (app == null) {
            return error.getErrorString();
        }

        if (error.getErrorCode() > 0) {
            return getString(R.string.api_error_http);
        }

        switch (error.getErrorCode()) {
            case API_OK:
                return getString(R.string.api_ok);
            case API_EINTERNAL:
                return getString(R.string.api_einternal);
            case API_EARGS:
                return getString(R.string.api_eargs);
            case API_EAGAIN:
                return getString(R.string.api_eagain);
            case API_ERATELIMIT:
                return getString(R.string.api_eratelimit);
            case API_EFAILED:
                return getString(R.string.api_efailed);
            case API_ETOOMANY:
                if (error.getErrorString().equals("Terms of Service breached")) {
                    return getString(R.string.api_etoomany_ec_download);
                } else if (error.getErrorString().equals("Too many concurrent connections or transfers")){
                    return getString(R.string.api_etoomay);
                } else {
                    return error.getErrorString();
                }
            case API_ERANGE:
                return getString(R.string.api_erange);
            case API_EEXPIRED:
                return getString(R.string.api_eexpired);
            case API_ENOENT:
                return getString(R.string.api_enoent);
            case API_ECIRCULAR:
                if (error.getErrorString().equals("Upload produces recursivity")) {
                    return getString(R.string.api_ecircular_ec_upload);
                } else if (error.getErrorString().equals("Circular linkage detected")){
                    return getString(R.string.api_ecircular);
                } else {
                    return error.getErrorString();
                }
            case API_EACCESS:
                return getString(R.string.api_eaccess);
            case API_EEXIST:
                return getString(R.string.api_eexist);
            case API_EINCOMPLETE:
                return getString(R.string.api_eincomplete);
            case API_EKEY:
                return getString(R.string.api_ekey);
            case API_ESID:
                return getString(R.string.api_esid);
            case API_EBLOCKED:
                if (error.getErrorString().equals("Not accessible due to ToS/AUP violation")) {
                    return getString(R.string.api_eblocked_ec_import_ec_download);
                } else if (error.getErrorString().equals("Blocked")) {
                    return getString(R.string.api_eblocked);
                } else {
                    return error.getErrorString();
                }
            case API_EOVERQUOTA:
                return getString(R.string.api_eoverquota);
            case API_ETEMPUNAVAIL:
                return getString(R.string.api_etempunavail);
            case API_ETOOMANYCONNECTIONS:
                return getString(R.string.api_etoomanyconnections);
            case API_EWRITE:
                return getString(R.string.api_ewrite);
            case API_EREAD:
                return getString(R.string.api_eread);
            case API_EAPPKEY:
                return getString(R.string.api_eappkey);
            case API_ESSL:
                return getString(R.string.api_essl);
            case API_EGOINGOVERQUOTA:
                return getString(R.string.api_egoingoverquota);
            case API_EMFAREQUIRED:
                return getString(R.string.api_emfarequired);
            case API_EMASTERONLY:
                return getString(R.string.api_emasteronly);
            case API_EBUSINESSPASTDUE:
                return getString(R.string.api_ebusinesspastdue);
            case PAYMENT_ECARD:
                return getString(R.string.payment_ecard);
            case PAYMENT_EBILLING:
                return getString(R.string.payment_ebilling);
            case PAYMENT_EFRAUD:
                return getString(R.string.payment_efraud);
            case PAYMENT_ETOOMANY:
                return getString(R.string.payment_etoomay);
            case PAYMENT_EBALANCE:
                return getString(R.string.payment_ebalance);
            case PAYMENT_EGENERIC:
            default:
                return getString(R.string.payment_egeneric_api_error_unknown);
        }
    }

    /**
     * Gets the translated string of an error received in a request.
     *
     * @param error MegaChatError received in the request
     * @return The translated string
     */
    public static String getTranslatedErrorString(MegaChatError error) {
        MegaApplication app = MegaApplication.getInstance();
        if (app == null) {
            return error.getErrorString();
        }

        switch (error.getErrorCode()) {
            case ERROR_OK:
                return getString(R.string.error_ok);
            case ERROR_ARGS:
                return getString(R.string.error_args);
            case ERROR_ACCESS:
                return getString(R.string.error_access);
            case ERROR_NOENT:
                return getString(R.string.error_noent);
            case ERROR_EXIST:
                return getString(R.string.error_exist);
            case ERROR_TOOMANY:
                return getString(R.string.error_toomany);
            case ERROR_UNKNOWN:
            default:
                return getString(R.string.error_unknown);
        }
    }
}

File: app/src/main/java/mega/privacy/android/app/utils/StringResourcesUtils.kt
package mega.privacy.android.app.utils

import android.content.Context
import mega.privacy.android.app.MegaApplication.Companion.getInstance
import mega.privacy.android.app.R
import nz.mega.sdk.MegaChatError
import nz.mega.sdk.MegaError

/**
 * Deprecated string resource util class
 */
object StringResourcesUtils {


    private val context: Context by lazy { getInstance() }

    /**
     * Gets the translated string of an error received in a request.
     *
     * @param error MegaError received in the request
     * @return The translated string
     */
    @JvmStatic
    @Deprecated(
        message = "MegaError should not be thrown in Presentation module, but instead mapped in Data module. For now however extension methods can be used",
        replaceWith = ReplaceWith("mega.privacy.android.app.presentation.extensions.MegaExceptionKt#getErrorStringId(MegaException)")
    )
    fun getTranslatedErrorString(error: MegaError): String {
        return getTranslatedErrorString(error.errorCode, error.errorString)
    }

    /**
     * Gets the translated string of a Mega error.
     *
     * @param errorCode MegaChat error code
     * @return The translated string
     */
    @JvmStatic
    @Deprecated(
        message = "MegaError should not be thrown in Presentation module, but instead mapped in Data module. For now however extension methods can be used",
        replaceWith = ReplaceWith("mega.privacy.android.app.presentation.extensions.MegaExceptionKt#getErrorStringId(MegaException)")
    )
    fun getTranslatedErrorString(errorCode: Int, errorString: String): String {
        return when (errorCode) {
            MegaError.API_OK -> context.getString(R.string.api_ok)
            MegaError.API_EINTERNAL -> context.getString(R.string.api_einternal)
            MegaError.API_EARGS -> context.getString(R.string.api_eargs)
            MegaError.API_EAGAIN -> context.getString(R.string.api_eagain)
            MegaError.API_ERATELIMIT -> context.getString(R.string.api_eratelimit)
            MegaError.API_EFAILED -> context.getString(R.string.api_efailed)
            MegaError.API_ETOOMANY -> when (errorString) {
                "Terms of Service breached" -> context.getString(R.string.api_etoomany_ec_download)
                "Too many concurrent connections or transfers" -> context.getString(R.string.api_etoomay)
                else -> errorString
            }
            MegaError.API_ERANGE -> context.getString(R.string.api_erange)
            MegaError.API_EEXPIRED -> context.getString(R.string.api_eexpired)
            MegaError.API_ENOENT -> context.getString(R.string.api_enoent)
            MegaError.API_ECIRCULAR -> when (errorString) {
                "Upload produces recursivity" -> context.getString(R.string.api_ecircular_ec_upload)
                "Circular linkage detected" -> context.getString(R.string.api_ecircular)
                else -> errorString
            }
            MegaError.API_EACCESS -> context.getString(R.string.api_eaccess)
            MegaError.API_EEXIST -> context.getString(R.string.api_eexist)
            MegaError.API_EINCOMPLETE -> context.getString(R.string.api_eincomplete)
            MegaError.API_EKEY -> context.getString(R.string.api_ekey)
            MegaError.API_ESID -> context.getString(R.string.api_esid)
            MegaError.API_EBLOCKED -> when (errorString) {
                "File removed as it violated our Terms of Service" -> context.getString(R.string.error_download_takendown_file)
                "Blocked" -> context.getString(R.string.api_eblocked)
                else -> errorString
            }
            MegaError.API_EOVERQUOTA -> context.getString(R.string.api_eoverquota)
            MegaError.API_ETEMPUNAVAIL -> context.getString(R.string.api_etempunavail)
            MegaError.API_ETOOMANYCONNECTIONS -> context.getString(R.string.api_etoomanyconnections)
            MegaError.API_EWRITE -> context.getString(R.string.api_ewrite)
            MegaError.API_EREAD -> context.getString(R.string.api_eread)
            MegaError.API_EAPPKEY -> context.getString(R.string.api_eappkey)
            MegaError.API_ESSL -> context.getString(R.string.api_essl)
            MegaError.API_EGOINGOVERQUOTA -> context.getString(R.string.api_egoingoverquota)
            MegaError.API_EMFAREQUIRED -> context.getString(R.string.api_emfarequired)
            MegaError.API_EMASTERONLY -> context.getString(R.string.api_emasteronly)
            MegaError.API_EBUSINESSPASTDUE -> context.getString(R.string.api_ebusinesspastdue)
            MegaError.PAYMENT_ECARD -> context.getString(R.string.payment_ecard)
            MegaError.PAYMENT_EBILLING -> context.getString(R.string.payment_ebilling)
            MegaError.PAYMENT_EFRAUD -> context.getString(R.string.payment_efraud)
            MegaError.PAYMENT_ETOOMANY -> context.getString(R.string.payment_etoomay)
            MegaError.PAYMENT_EBALANCE -> context.getString(R.string.payment_ebalance)
            MegaError.PAYMENT_EGENERIC -> if (errorCode > 0) {
                context.getString(R.string.api_error_http)
            } else {
                context.getString(R.string.payment_egeneric_api_error_unknown)
            }
            else -> if (errorCode > 0) {
                context.getString(R.string.api_error_http)
            } else {
                context.getString(R.string.payment_egeneric_api_error_unknown)
            }
        }
    }

    /**
     * Gets the translated string of an error received in a request.
     *
     * @param error MegaChatError received in the request
     * @return The translated string
     */
    @JvmStatic
    @Deprecated(
        message = "MegaChatError should not be thrown in Presentation module, but instead mapped in Data module. For now however extension methods can be used",
        replaceWith = ReplaceWith("mega.privacy.android.app.presentation.extensions.MegaExceptionKt#getChatErrorStringId(MegaException)")
    )
    fun getTranslatedErrorString(error: MegaChatError): String {
        return getTranslatedChatErrorString(error.errorCode)
    }

    /**
     * Gets the translated string of a MegaChat error code.
     *
     * @param errorCode MegaChat error code
     * @return The translated string
     */
    @JvmStatic
    @Deprecated(
        message = "MegaChatError should not be thrown in Presentation module, but instead mapped in Data module. For now however extension methods can be used",
        replaceWith = ReplaceWith("mega.privacy.android.app.presentation.extensions.MegaExceptionKt#getChatErrorStringId(MegaException)")
    )
    fun getTranslatedChatErrorString(errorCode: Int): String {
        return when (errorCode) {
            MegaChatError.ERROR_OK -> context.getString(R.string.error_ok)
            MegaChatError.ERROR_ARGS -> context.getString(R.string.error_args)
            MegaChatError.ERROR_ACCESS -> context.getString(R.string.error_access)
            MegaChatError.ERROR_NOENT -> context.getString(R.string.error_noent)
            MegaChatError.ERROR_EXIST -> context.getString(R.string.error_exist)
            MegaChatError.ERROR_TOOMANY -> context.getString(R.string.error_toomany)
            MegaChatError.ERROR_UNKNOWN -> context.getString(R.string.error_unknown)
            else -> context.getString(R.string.error_unknown)
        }
    }
}